microovn/ovn/dpu/dpu.go

```diff
@@ -0,0 +1,99 @@
+package dpu
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	"github.com/canonical/lxd/shared"
+	"github.com/canonical/lxd/shared/logger"
+	"github.com/canonical/microcluster/v2/state"
+
+	ovnCmd "github.com/canonical/microovn/microovn/ovn/cmd"
+)
+
+// detectDPU uses lspci to find if microovn is running on a DPU and if so returns
+// the PCI address
+func detectDPU(ctx context.Context) (string, error) {
+	out, err := shared.RunCommandContext(ctx, "lspci")
+	if err != nil {
+		return "", fmt.Errorf("failed to run lspci: %w", err)
+	}
+
+	for _, line := range strings.Split(out, "\n") {
+		if strings.Contains(line, "Mellanox") &&
+			(strings.Contains(line, "BlueField") ||
+				strings.Contains(line, "ConnectX")) {
+
+			fields := strings.Fields(line)
+			if len(fields) > 0 {
+				pci := fields[0]
+				logger.Infof("DPU detected at PCI %s", pci)
+				return pci, nil
+			}
+		}
+	}
+
+	return "", nil
+}
+
+// getSerialNumber parses lspci and gets the serial number for the DPU, given a
+// specific PCI address.
+func getDPUSerial(ctx context.Context, pci string) (string, error) {
+	out, err := shared.RunCommandContext(ctx, "lspci", "-vvv", "-s", pci)
+	if err != nil {
+		return "", fmt.Errorf("failed to run lspci -vvv -s %s: %w", pci, err)
+	}
+
+	for _, line := range strings.Split(out, "\n") {
+		line = strings.TrimSpace(line)
+		if strings.HasPrefix(line, "[SN] Serial number:") {
+			serial := strings.SplitN(line, ":", 2)[1]
+			logger.Infof("DPU serial number detected: %s", serial)
+			return serial, nil
+		}
+	}
+
+	return "", fmt.Errorf("serial number not found for DPU at PCI %s", pci)
+}
+
+// setDPUSerial writes the external-id the ovn-cms-options.
+func setOVNDPUSerial(ctx context.Context, s state.State, serial string) error {
+	_, err := ovnCmd.VSCtl(
+		ctx,
+		s,
+		"set", "Open_vSwitch", ".",
+		fmt.Sprintf("external-ids:ovn-cms-options=card-serial-number=%s", serial),
```

Comment:
The `external-ids:ovn-cms-options` is a string that contains a comma separated list of key/value pairs [0], and may be populated by multiple entities running on the same system.

This approach would overwrite any existing values already set by other systems.

0: https://manpages.ubuntu.com/manpages/questing/man8/ovn-controller.8.html